core/lib/Cardano/Address/Style/Byron.hs

```diff
@@ -326,47 +332,55 @@ prettyErrInspectAddress = \case
+-- 'ErrInspectAddress' if the address isn't a byron address.
 --
 -- @since 2.0.0
 inspectByronAddress :: forall m. MonadThrow m => Maybe XPub -> Address -> m Json.Value
 inspectByronAddress = inspectAddress
 {-# DEPRECATED inspectByronAddress "use qualified 'inspectAddress' instead." #-}
 
--- | Analyze an 'Address' to know whether it's a Byron address or not.
--- Throws 'ByronAddrError' if the address isn't a byron address, or return a
--- structured JSON that gives information about an address.
+-- | Determines whether an 'Address' is a Byron address.
+--
+-- Returns a JSON object with information about the address, or throws
+-- 'ErrInspectAddress' if the address isn't a byron address.
 --
 -- @since 3.0.0
 inspectAddress :: forall m. MonadThrow m => Maybe XPub -> Address -> m Json.Value
-inspectAddress mRootPub addr = do
-    payload <- either (throwM . DeserialiseError) pure
-        $ CBOR.deserialiseCbor CBOR.decodeAddressPayload bytes
+inspectAddress mRootPub addr = either throwM (pure . toJSON) $
+    eitherInspectAddress mRootPub addr
+
+-- | Determines whether an 'Address' is a Byron address.
+--
+-- Returns either details about the 'Address', or 'ErrInspectAddress' if it's
+-- not a valid address.
+--
+-- @since 3.4.0
+eitherInspectAddress :: Maybe XPub -> Address -> Either ErrInspectAddress AddressInfo
+eitherInspectAddress mRootPub addr = do
+    payload <- first DeserialiseError $
+        CBOR.deserialiseCbor CBOR.decodeAddressPayload bytes
 
-    (root, attrs) <- either (throwM . DeserialiseError) pure
-        $ CBOR.deserialiseCbor decodePayload payload
+    (root, attrs) <- first DeserialiseError $
+        CBOR.deserialiseCbor decodePayload payload
 
     path <- do
-        attr <- maybe (throwM MissingExpectedDerivationPath) pure
-            (find ((== 1) . fst) attrs)
+        attr <- maybe (Left MissingExpectedDerivationPath) Right $
+            find ((== 1) . fst) attrs
         case mRootPub of
-            Nothing ->
-                pure $ toJSON $ T.unpack $ T.decodeUtf8 $ encode EBase16 $ snd attr
-            Just rootPub ->
-                decryptPath rootPub attr
-
-    ntwrk <- either (throwM . DeserialiseError) pure
-        $ CBOR.deserialiseCbor CBOR.decodeProtocolMagicAttr payload
-
-    pure $ Json.object
-        [ "address_style"   .= Json.String "Byron"
-        , "stake_reference" .= Json.String "none"
-        , "address_root"    .= T.unpack (T.decodeUtf8 $ encode EBase16 root)
-        , "derivation_path" .= path
-        , "network_tag"     .= maybe Json.Null toJSON ntwrk
-        ]
+            Nothing -> Right $ EncryptedDerivationPath $ snd attr
+            Just rootPub -> decryptPath attr rootPub
+
+    ntwrk <- bimap DeserialiseError (fmap NetworkTag) $
+        CBOR.deserialiseCbor CBOR.decodeProtocolMagicAttr payload
+
+    pure AddressInfo
+        { infoAddressRoot = root
+        , infoPayload = path
+        , infoNetworkTag = ntwrk
+        }
```

Comment:
Arguably, that should be the _proper_ type for `Address` already instead of a wrapper around `ByteString`